README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# mmcards

Games of chance serve as essential learning tools for understanding
probability and statistics. The ‘mmcards’ package, short for Mighty
Metrika cards, offers utility functions for handling playing cards,
streamlining the development of educational Shiny applications related
to games of chance.

## Installation

You can install the released version of ‘mmcards’ from
[CRAN](https://CRAN.R-project.org):

``` r
install.packages("mmcards")
```

To install the development version of ‘mmcards’, you can use the
[devtools](https://devtools.r-lib.org/) package:

``` r
# install.packages("devtools")
devtools::install_github("mightymetrika/mmcards")
```

## Who Has the Highest Card: A Basic Example

This section demonstrates a simple game called “Who Has the Highest
Card,” which shows you how to use mmcards to create a shuffled standard
deck and deal cards to players.

``` r
library(mmcards)

# Define the Players
p1 <- NULL
p2 <- NULL

# Create a shuffled standard deck of 52 cards using a random number seed for reproducibility
game_deck <- shuffle_deck(seed = 147)

# Deal a card to each player
game_deck <- deal_card(game_deck)
p1 <- game_deck$dealt_card

game_deck <- deal_card(game_deck)
p2 <- game_deck$dealt_card

# Display the cards and determine the winner
paste("p1 has:", p1$card, "with a value of", p1$value)
#> [1] "p1 has: 5C with a value of 5"
paste("p2 has:", p2$card, "with a value of", p2$value)
#> [1] "p2 has: KH with a value of 13.5"
paste0("The winner is: ", ifelse(p1$value > p2$value, "p1", "p2"))
#> [1] "The winner is: p2"
```

## Alternative Decks

To accommodate various use-cases, shuffle_deck() allows users to create
specialized decks like AnonymousDeck, InterleavedDeck, or PairedDeck.
You can learn more about these decks in the “alternative-decks”
vignette.

## Adding Card Images

Use the i_deck() function to add images to your card decks. You’ll need
to have the images stored on your computer.

The following example was used to create a Shiny application which flips
through a deck of shuffled cards showing the card image and the card
value.

In this example, my working directory was set to
“C:/Users/Administrator/Desktop/mmcards” and a deck of PNG
[vector-playing-cards](https://code.google.com/archive/p/vector-playing-cards/downloads)
was stored at
“C:/Users/Administrator/Desktop/PNG-cards-1.3/PNG-cards-1.3”.

``` r
#install.packages("shiny")
library(shiny)

    # Initialize a shuffled and image-embedded deck
    ic <- i_deck(deck = shuffle_deck(seed = 42),
                 i_path = paste0(gsub("mmcards", "", getwd()),
                                 "PNG-cards-1.3/PNG-cards-1.3"))

    # UI
    ui <- fluidPage(
      titlePanel("Card Deck Shiny App"),

      fluidRow(
        column(6,
               # Image output
               imageOutput("cardImage", height = "auto")
        ),
        column(6,
               # Text output
               textOutput("cardValue"),

               # Deal card button
               actionButton("deal_card", "Deal Card")
        )
      )
    )

    # Server logic
    server <- function(input, output, session) {

      # Initialize reactive variable to store current deck
      current_deck <- reactiveVal(ic)

      # Output for card image
      output$cardImage <- renderImage({
        # Initialize or get the current deck
        deck <- current_deck()

        # Return image path
        list(src = deck$icard[1],
             width = 250,
             height = 350)
      }, deleteFile = FALSE)

      # Output for card value
      output$cardValue <- renderText({
        # Initialize or get the current deck
        deck <- current_deck()

        # Return card value
        return(deck$value[1])
      })

      # Deal card logic
      observeEvent(input$deal_card, {
        # Get the current deck
        deck <- current_deck()

        # Deal a card
        new_deck <- deal_card(deck)

        # Update the current deck
        current_deck(new_deck$updated_deck)
      })
    }

    # Run the app
    shinyApp(ui, server)
    
```


NEWS.md
# mmcards (development version)

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(mmcards)

test_check("mmcards")


tests/testthat/test-deck_funcs.R
test_that("standard_deck works with defaults", {
  deck <- standard_deck()

  expect_equal(nrow(deck), 52)
  expect_s3_class(deck, "data.frame")
  expect_s3_class(deck, "StandardDeck")
})

test_that("shuffle_deck works with standard_deck", {
  #Get deck
  sdeck <- shuffle_deck(seed = 100)

  # Test expectations
  expect_equal(nrow(sdeck), 52)
  expect_s3_class(sdeck, "data.frame")
  expect_s3_class(sdeck, "ShuffledDeck")
  expect_s3_class(sdeck, "StandardDeck")
})

test_that("shuffle_deck works with anonymous function", {

  #Get deck
  sdeck <- shuffle_deck(deck_of_cards = function(x){stats::runif(52, 1, 52)},
                        seed = 100)

  # Test expectations
  expect_equal(nrow(sdeck), 52)
  expect_s3_class(sdeck, "data.frame")
  expect_s3_class(sdeck, "ShuffledDeck")
  expect_s3_class(sdeck, "AnonymousDeck")
})

test_that("shuffle_deck works for interleaved decks", {

  #Get deck
  sdeck <- shuffle_deck(deck_of_cards = function(x) {list(stats::rnorm(26, 1, 2),
                                                          stats::rnorm(26, 1.5, 1.5))},
                        seed = 100)

  # Test expectations
  expect_equal(nrow(sdeck), 52)
  expect_s3_class(sdeck, "data.frame")
  expect_s3_class(sdeck, "ShuffledDeck")
  expect_s3_class(sdeck, "InterleavedDeck")
})

test_that("shuffle_deck works for interleaved decks with paired = TRUE", {

  #Get deck
  sdeck <- shuffle_deck(deck_of_cards = function(x) {list(c(1:26),
                                                          c(1:26))},
                        seed = 100,
                        paired = TRUE)

  # Test expectations
  expect_equal(nrow(sdeck), 52)
  expect_s3_class(sdeck, "data.frame")
  expect_s3_class(sdeck, "ShuffledDeck")
  expect_s3_class(sdeck, "InterleavedDeck")

  # Test that pairing worked
  expect_equal(sdeck$value[[1]],sdeck$value[[2]])
  expect_equal(sdeck$value[[11]],sdeck$value[[12]])
  expect_equal(sdeck$value[[21]],sdeck$value[[22]])
  expect_equal(sdeck$value[[31]],sdeck$value[[32]])
  expect_equal(sdeck$value[[41]],sdeck$value[[42]])
  expect_equal(sdeck$value[[51]],sdeck$value[[52]])
})

test_that("deal_card works with standard deck", {

  # Get deck
  sdeck <- shuffle_deck(seed = 100)

  # Deal one card
  udeck <- deal_card(sdeck)

  # Test expectations
  expect_equal(nrow(udeck$dealt_card), 1)
  expect_equal(nrow(udeck$updated_deck), 51)
  expect_s3_class(udeck, "UpDeck")

  # Deal second card
  udeck <- deal_card(udeck$updated_deck)

  # Test expectations
  expect_equal(nrow(udeck$dealt_card), 1)
  expect_equal(nrow(udeck$updated_deck), 50)
  expect_s3_class(udeck, "UpDeck")

})

test_that("deal_card works with standard deck and up_deck class", {

  # Get deck
  sdeck <- shuffle_deck(seed = 100)

  # Deal one card
  udeck <- deal_card(sdeck)

  # Deal second card
  udeck <- deal_card(udeck)

  # Test expectations
  expect_equal(nrow(udeck$dealt_card), 1)
  expect_equal(nrow(udeck$updated_deck), 50)
  expect_s3_class(udeck, "UpDeck")

})



tests/testthat/test-image_funcs.R
test_that("i_deck adds images to a deck of cards", {

  ic <- i_deck(deck = standard_deck(),
               i_path = getwd())

  expect_s3_class(ic, "ImgDeck")
  expect_s3_class(ic, "StandardDeck")

})

# test_that("i_deck adds images to a deck of cards", {
#
#   ic <- i_deck(deck = standard_deck(),
#                 i_path = paste0(gsub("mmcards", "", getwd()),
#                                 "PNG-cards-1.3/PNG-cards-1.3"))
#
#   expect_s3_class(ic, "ImgDeck")
#   expect_equal(2 * 2, 4)
# })
#
# test_that("i_deck works with a Shiny application", {
#
#     library(shiny)
#
#     # Initialize a shuffled and image-embedded deck
#     ic <- i_deck(deck = shuffle_deck(seed = 42),
#                  i_path = paste0(gsub("mmcards", "", getwd()),
#                                  "PNG-cards-1.3/PNG-cards-1.3"))
#
#     # UI
#     ui <- fluidPage(
#       titlePanel("Card Deck Shiny App"),
#
#       fluidRow(
#         column(6,
#                # Image output
#                imageOutput("cardImage", height = "auto")
#         ),
#         column(6,
#                # Text output
#                textOutput("cardValue"),
#
#                # Deal card button
#                actionButton("deal_card", "Deal Card")
#         )
#       )
#     )
#
#     # Server logic
#     server <- function(input, output, session) {
#
#       # Initialize reactive variable to store current deck
#       current_deck <- reactiveVal(ic)
#
#       # Output for card image
#       output$cardImage <- renderImage({
#         # Initialize or get the current deck
#         deck <- current_deck()
#
#         # Return image path
#         list(src = deck$icard[1],
#              width = 250,
#              height = 350)
#       }, deleteFile = FALSE)
#
#       # Output for card value
#       output$cardValue <- renderText({
#         # Initialize or get the current deck
#         deck <- current_deck()
#
#         # Return card value
#         return(deck$value[1])
#       })
#
#       # Deal card logic
#       observeEvent(input$deal_card, {
#         # Get the current deck
#         deck <- current_deck()
#
#         # Deal a card
#         new_deck <- deal_card(deck)
#
#         # Update the current deck
#         current_deck(new_deck$updated_deck)
#       })
#     }
#
#     # Run the app
#     app <- shinyApp(ui, server)
#
#     expect_s3_class(app, "shiny.appobj")
# })


R/image_funcs.R
#' Add image paths to a deck of cards
#'
#' This function takes a deck of cards and adds image paths to each card.
#' It produces a new deck that includes the original information along with the
#' image paths. The function is designed to work with various types of decks and
#' allows for customization of image paths.
#'
#' @param deck A data frame representing a deck of cards. Default is standard_deck(),
#' but the function is also optimized to work with shuffled_deck().
#' @param i_path The file path to the folder containing the card images.
#' @param cards A vector of card names corresponding to the images. Default includes
#' all cards based on a deck created with standard_deck()
#' @param i_names A vector of image file names corresponding to the cards.
#' Should be in the same order as `cards` the default is vector_playing_cards
#' which works when `deck` is set to standard_deck() and `cards` uses the default
#' vector.
#' @param i_type The file extension of the image files (e.g., "png", "jpg").
#'
#' @return A data frame that includes the original deck along with image paths for
#' each card. Inherits the class "ImgDeck" in addition to the original deck's classes.
#'
#' @examples
#' #The following example shows how to add image paths to the 'mmcards' default
#' #StandardDeck when the images referenced in ?vector_playing_cards are stored
#' #in the working directory.
#'
#'  image_deck <- i_deck(deck = standard_deck(),
#'                       i_path = getwd())
#'
#'  head(image_deck)
#'
#'  #See the README for an example of how to use i_deck to incorporate images
#'  #into a 'shiny' application.
#'
#' @export
i_deck <- function(deck,
                    i_path,
                    cards = c("2C", "2D", "2H", "2S", "3C", "3D", "3H", "3S",
                              "4C", "4D", "4H", "4S", "5C", "5D", "5H", "5S",
                              "6C", "6D", "6H", "6S", "7C", "7D", "7H", "7S",
                              "8C", "8D", "8H", "8S", "9C", "9D", "9H", "9S",
                              "10C", "10D", "10H", "10S", "JC", "JD", "JH", "JS",
                              "QC", "QD", "QH", "QS", "KC", "KD", "KH", "KS",
                              "AC", "AD", "AH", "AS"),
                    i_names = vector_playing_cards,
                    i_type = "png"){

  # Check if the input deck inherits from "StandardDeck" or "ShuffledDeck"
  if (!inherits(deck, c("StandardDeck", "ShuffledDeck"))) {
    stop("The 'deck' must inherit from 'StandardDeck' or 'ShuffledDeck'.")
  }

  # Check if 'cards' and 'i_names' have the same length
  if (length(cards) != length(i_names)) {
    stop("The vectors 'cards' and 'i_names' must have the same length.")
  }

  # Check if each value in 'cards' can be found in the 'card' column of 'deck'
  if (!all(cards %in% deck$card)) {
    stop("All values in 'cards' must correspond to a card in the 'deck'.")
  }

  # Get folder with images
  image_paths <- file.path(i_path, paste0(i_names, ".", i_type))

  # Create image data frame
  image_frame <- data.frame(card = cards,
                            icard = image_paths)
  # Store the classes of deck
  deck_classes <- class(deck)

  # Merge deck and image_frame without sorting
  ideck <- merge(deck, image_frame, by = "card", sort = FALSE)

  # To ensure the order is exactly the same as the original deck
  ideck <- ideck[match(deck$card, ideck$card),]

  #Add class "ideck"
  class(ideck) <- append("ImgDeck", deck_classes)

  # Return
  return(ideck)
}


data-raw/vector_playing_cards.R
## code to prepare `vector_playing_cards` dataset goes here

vector_playing_cards <- c("2_of_clubs", "2_of_diamonds", "2_of_hearts", "2_of_spades",
                          "3_of_clubs", "3_of_diamonds", "3_of_hearts", "3_of_spades",
                          "4_of_clubs", "4_of_diamonds", "4_of_hearts", "4_of_spades",
                          "5_of_clubs", "5_of_diamonds", "5_of_hearts", "5_of_spades",
                          "6_of_clubs", "6_of_diamonds", "6_of_hearts", "6_of_spades",
                          "7_of_clubs", "7_of_diamonds", "7_of_hearts", "7_of_spades",
                          "8_of_clubs", "8_of_diamonds", "8_of_hearts", "8_of_spades",
                          "9_of_clubs", "9_of_diamonds", "9_of_hearts", "9_of_spades",
                          "10_of_clubs", "10_of_diamonds", "10_of_hearts", "10_of_spades",
                          "jack_of_clubs", "jack_of_diamonds", "jack_of_hearts", "jack_of_spades",
                          "queen_of_clubs", "queen_of_diamonds", "queen_of_hearts", "queen_of_spades",
                          "king_of_clubs", "king_of_diamonds", "king_of_hearts", "king_of_spades",
                          "ace_of_clubs", "ace_of_diamonds", "ace_of_hearts", "ace_of_spades")

usethis::use_data(vector_playing_cards, overwrite = TRUE)


R/deck_funcs.R
#' Generate a Standard Deck of Playing Cards
#'
#' This function creates a standard deck of playing cards represented as a data
#' frame. The deck includes suits, ranks, and values for each card.
#'
#' @param suits A character vector specifying the suits in the deck. Default is
#' c('C', 'D', 'H', 'S') for Clubs, Diamonds, Hearts, and Spades.
#' @param ranks A character vector specifying the ranks in the deck. Default is
#' c('2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A') for ranks
#' 2 to Ace.
#' @param values A numeric vector specifying the values assigned to each card in
#' the deck. Default is a sequence from 2 to 14.75 incremented by 0.25.
#'
#' @return A data frame representing the deck of cards. The data frame has four
#' columns: `rank`, `suit`, `card`, and `value`. The data frame also has class
#' attributes "StandardDeck" and "data.frame".
#'
#' @examples
#' deck <- standard_deck()
#' head(deck)
#' tail(deck)
#'
#' @export
standard_deck <- function(suits = c('C', 'D', 'H', 'S'),
                          ranks = c('2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A'),
                          values = seq(2, 14.75, by = 0.25)){

  # Check if suits and ranks are character vectors
  if (!is.character(suits)) {
    stop("The 'suits' argument must be a character vector.")
  }

  if (!is.character(ranks)) {
    stop("The 'ranks' argument must be a character vector.")
  }

  # Check if the length of suits is 4
  if (length(suits) != 4) {
    stop("The 'suits' argument must have a length of 4.")
  }

  # Check if values is a numeric vector
  if (!is.numeric(values)) {
    stop("The 'values' argument must be a numeric vector.")
  }

  # Check if the length of values matches the number of cards in the deck
  if (length(values) != length(suits) * length(ranks)) {
    stop("The length of the 'values' argument must match the number of cards in the deck (length of 'suits' * length of 'ranks').")
  }

  # Create the deck
  deck <- expand.grid(rank = ranks, suit = suits)
  deck$card <- paste0(deck$rank, deck$suit)

  # Turn suit to factor
  deck$suit <- factor(deck$suit,
                      levels = suits,
                      labels = suits)

  # Order deck
  deck <- deck[order(deck$rank, deck$suit), ]

  # Add values to cards
  deck$value <- values

  # Turn card to factor using the current order
  deck$card <- factor(deck$card, levels = deck$card)

  #Set class
  class(deck) <- c("StandardDeck", "data.frame")

  return(deck)
}


#' Shuffle a Deck of Cards
#'
#' This function shuffles a deck of cards and returns the shuffled deck.
#' The function can handle standard decks, anonymous decks, and interleaved decks.
#' For interleaved decks, an option to pair shuffle is also available.
#'
#' @param deck_of_cards An anonymous function that returns a deck of cards as
#' either a data frame or a list of two numeric vectors for interleaved decks.
#' Default is `function(x){standard_deck()}`.
#' @param seed An optional seed for reproducibility. Default is NULL.
#' @param paired Logical flag to indicate if the interleaved deck should be pair
#' shuffled. Default is FALSE.
#'
#' @return A data frame representing the shuffled deck of cards. The data frame
#' inherits various classes based on its type. All shuffled decks will have the
#' classes "ShuffledDeck" and "data.frame". Additional class inheritance depends
#' on the `deck_of_cards` parameter:
#' - "StandardDeck" if `deck_of_cards` returns a standard deck (default)
#' - "AnonymousDeck" if `deck_of_cards` returns a single vector
#' - "InterleavedDeck" if `deck_of_cards` returns a list of two vectors.
#' If the `paired` parameter is set to TRUE, an interleaved deck will also
#' inherit the class "PairedDeck".
#'
#' @examples
#' # Standard deck
#' std_deck <- shuffle_deck()
#' head(std_deck)
#'
#' # Anonymous deck
#' anon_deck <- shuffle_deck(deck_of_cards = function(x){runif(52, 1, 10)})
#' head(anon_deck)
#'
#' # Interleaved deck
#' interleaved_deck <- shuffle_deck(
#'                         deck_of_cards = function(x){list(runif(26, 1, 5),
#'                                                          runif(26, 6, 10))})
#' head(interleaved_deck)
#'
#' # Paired interleaved deck
#' paired_deck <- shuffle_deck(
#'                   deck_of_cards = function(x){list(runif(26, 1, 5),
#'                                                    runif(26, 6, 10))},
#'                   paired = TRUE)
#' head(paired_deck)
#'
#' @export
shuffle_deck <- function(deck_of_cards = function(x){standard_deck()}, seed = NULL,
                         paired = FALSE) {

  # Check if 'seed' is a valid single integer
  if (!is.null(seed) && (!is.numeric(seed) || length(seed) != 1)) {
    stop("The 'seed' parameter must be a single integer.")
  }

  # Check if 'paired' is a logical value
  if (!is.logical(paired) || length(paired) != 1) {
    stop("The 'paired' parameter must be a single logical value.")
  }

  # Set seed for reproducibility
  if (!is.null(seed)) {
    set.seed(seed)
  }

  # Generate the values based on the anonymous function
  decks <- deck_of_cards(NULL)

  # Check if the result from 'deck_of_cards' is valid
  if (!(inherits(decks, "StandardDeck") ||
        (is.numeric(decks) && is.vector(decks)) ||
        (is.list(decks) && length(decks) == 2 &&
         is.numeric(decks[[1]]) && is.numeric(decks[[2]])))) {
    stop("The 'deck_of_cards' function must return either a 'StandardDeck', a numeric vector, or a list of two numeric vectors.")
  }

  # If the result is a list with two elements, treat them as separate decks
  if (is.list(decks) && length(decks) == 2) {

    # Check if the two decks are of the same length
    if (length(decks[[1]]) != length(decks[[2]])) {
      stop("Both decks must be of the same length for interleaving!")
      }

    # Create separate decks for player and computer
    A_deck <- data.frame(card = paste("A", 1:length(decks[[1]]), sep = "_"), value = decks[[1]])
    B_deck <- data.frame(card = paste("B", 1:length(decks[[2]]), sep = "_"), value = decks[[2]])

    # Shuffle each deck
    if (paired == TRUE){
      index <- sample(nrow(A_deck))

      A_deck <- A_deck[index, ]
      B_deck <- B_deck[index, ]
    } else {
      A_deck <- A_deck[sample(nrow(A_deck)), ]
      B_deck <- B_deck[sample(nrow(B_deck)), ]
    }

    # Interleave the decks
    interleaved_deck <- rbind(A_deck, B_deck)[order(rep(1:nrow(A_deck), 2)), ]

    # Add classes and return deck
    class(interleaved_deck) <- c("InterleavedDeck", "ShuffledDeck", "data.frame")
    if (paired == TRUE){
      class(interleaved_deck) <- append("PairedDeck", class(interleaved_deck))
    }

    return(interleaved_deck)

    } else if (!inherits(decks, "StandardDeck")) {

      # If it's just one deck, treat it as before
      values <- decks
      ordered_indices <- order(values, decreasing = FALSE)
      ordered_deck <- data.frame(card = 1:length(values), value = values[ordered_indices])

      # Shuffle the ordered deck
      shuffled_deck <- ordered_deck[sample(nrow(ordered_deck)), ]

      # Add classes and return deck
      class(shuffled_deck) <- c("AnonymousDeck", "ShuffledDeck", "data.frame")
      return(shuffled_deck)

      } else {

        # Shuffle the provided deck
        shuffled_deck <- decks[sample(nrow(decks)), ]

        # Add classes and return deck
        class(shuffled_deck) <- c("StandardDeck", "ShuffledDeck", "data.frame")
        return(shuffled_deck)
      }
}

#' Deal a Card from the Deck
#'
#' This function deals the top card from a given deck and returns the dealt card
#' along with the updated deck.
#'
#' @param current_deck A data frame representing the current deck of cards. This
#' can either be a standard deck, an anonymous deck, or an interleaved deck.
#' The function also accepts an object of class "UpDeck" which contains an
#' updated deck and the last dealt card.
#'
#' @return A list containing two elements: `dealt_card`, a data frame representing
#' the card that was dealt, and `updated_deck`, a data frame representing the
#' remaining cards in the deck. The list has the class attribute "UpDeck".
#'
#' @examples
#' # Using a standard deck
#' std_deck <- standard_deck()
#' result <- deal_card(std_deck)
#' result$dealt_card
#' result$updated_deck
#'
#' # Using an "UpDeck" object
#' result2 <- deal_card(result)
#' result2$dealt_card
#' result2$updated_deck
#'
#' @export
deal_card <- function(current_deck) {

  # Check if 'current_deck' inherits from one of the expected classes
  if (!inherits(current_deck, c("StandardDeck", "ShuffledDeck", "UpDeck"))) {
    stop("The 'current_deck' must inherit from 'StandardDeck', 'ShuffledDeck', or 'UpDeck'.")
  }


  if (!inherits(current_deck, "UpDeck")){

    if (nrow(current_deck) == 0) {
      stop("No more cards in the deck!")
    }

    card_to_deal <- current_deck[1, ]
    updated_deck <- current_deck[-1, ]  # Remove the top card

  } else {

    if (nrow(current_deck$updated_deck) == 0) {
      stop("No more cards in the deck!")
    }

    card_to_deal <- current_deck$updated_deck[1, ]
    updated_deck <- current_deck$updated_deck[-1, ]  # Remove the top card

  }

  up_deck <- list(dealt_card = card_to_deal, updated_deck = updated_deck)
  class(up_deck) <- "UpDeck"

  return(up_deck)
}
